File: go/solutions/0350-intersection-of-two-arrays-ii.go
package solutions

func Intersect(nums1 []int, nums2 []int) []int {
	output := make([]int, 0)
	set := make(map[int]int)

	minArr, maxArr := nums1, nums2
	if len(nums1) > len(nums2) {
		minArr, maxArr = nums2, nums1
	}

	for _, num := range minArr {
		set[num]++
	}

	for _, num := range maxArr {
		if set[num] != 0 {
			output = append(output, num)
			set[num]--
		}
	}

	return output
}

File: js/0350-intersection-of-two-arrays-ii.js
/**
 * @param {number[]} nums1
 * @param {number[]} nums2
 * @return {number[]}
 */
var intersect = function (nums1, nums2) {
  var minArr = nums1,
    maxArr = nums2;

  if (nums1.length > nums2.length) {
    (maxArr = nums1), (minArr = nums2);
  }

  return minArr.filter((val) => {
    var res = maxArr.includes(val);

    if (res) {
      maxArr.splice(maxArr.indexOf(val), 1);
    }

    return res;
  });
};

module.exports = { intersect };
